File: medium/TreePlucking.java
import java.io.*;
import java.util.*;

public class TreePlucking {
    static class Node {
        long val;
        int deg;
        List<Integer> adj;
        Node() { adj = new ArrayList<>(); deg = 0; val = 0; }
    }
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        Node[] nodes = new Node[n];
        for (int i = 0; i < n; i++) nodes[i] = new Node();
        for (int i = 0; i < n; i++) {
            nodes[i].val = Long.parseLong(br.readLine().trim());
        }
        for (int i = 0; i < n - 1; i++) {
            String[] parts = br.readLine().trim().split("\\s+");
            int u = Integer.parseInt(parts[0]) - 1;
            int v = Integer.parseInt(parts[1]) - 1;
            nodes[u].adj.add(v);
            nodes[v].adj.add(u);
            nodes[u].deg++;
            nodes[v].deg++;
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingLong(a -> nodes[a[0]].val));
        boolean[] removed = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (nodes[i].deg <= 1) {
                pq.offer(new int[]{i, i});
            }
        }
        long totalCost = 0;
        int removedCount = 0;
        while (removedCount < n) {
            int[] curr = pq.poll();
            int u = curr[0];
            if (removed[u]) continue;
            totalCost += nodes[u].val;
            removed[u] = true;
            removedCount++;
            for (int w : nodes[u].adj) {
                if (!removed[w]) {
                    nodes[w].val += nodes[u].val;
                    nodes[w].deg--;
                    if (nodes[w].deg == 1) {
                        pq.offer(new int[]{w, w});
                    }
                }
            }
        }
        System.out.println(totalCost);
    }
}

File: medium/TreePlucking.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    long long val;
    int deg;
    vector<int> adj;
    Node() { val = 0; deg = 0; }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if(!(cin >> n)) return 0;
    vector<Node> nodes(n);
    for(int i = 0; i < n; i++) cin >> nodes[i].val;
    for(int i = 0; i < n - 1; i++) {
        int u, v;
        cin >> u >> v;
        u--; v--;
        nodes[u].adj.push_back(v);
        nodes[v].adj.push_back(u);
        nodes[u].deg++;
        nodes[v].deg++;
    }
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> pq;
    vector<bool> removed(n, false);
    for(int i = 0; i < n; i++) {
        if(nodes[i].deg <= 1) {
            pq.push({nodes[i].val, i});
        }
    }
    long long totalCost = 0;
    int removedCount = 0;
    while(removedCount < n) {
        auto [val, u] = pq.top();
        pq.pop();
        if(removed[u]) continue;
        totalCost += nodes[u].val;
        removed[u] = true;
        removedCount++;
        for(int w : nodes[u].adj) {
            if(!removed[w]) {
                nodes[w].val += nodes[u].val;
                nodes[w].deg--;
                if(nodes[w].deg == 1) {
                    pq.push({nodes[w].val, w});
                }
            }
        }
    }
    cout << totalCost << "\n";
    return 0;
}
